File: clients/apple/MultiLiveTV/Services/VodMergeService.swift
import Foundation

enum VodMergeService {
    private static let maxVariants = 16

    struct MergedVodItem {
        let item: VodItemRaw
        let variants: [VodVariant]
        let primarySourceId: Int
    }

    static func isCompatibleVodMatch(_ primary: VodItemRaw, _ candidate: VodItemRaw) -> Bool {
        let primaryTitle = HomeFeed.normalizeTitle(primary.vodName)
        let candidateTitle = HomeFeed.normalizeTitle(candidate.vodName)
        guard primaryTitle == candidateTitle, !primaryTitle.isEmpty else { return false }
        let primaryYear = HomeFeed.normalizeYear(primary.vodYear)
        let candidateYear = HomeFeed.normalizeYear(candidate.vodYear)
        if primaryYear.isEmpty || candidateYear.isEmpty { return true }
        return primaryYear == candidateYear
    }

    static func mergeVodItems(_ items: [MergeableVodItem], store: SourceStore) -> [MergedVodItem] {
        let sourceOrder = Dictionary(uniqueKeysWithValues: store.enabled().enumerated().map { ($1.id, $0) })
        let (groups, order) = groupMergeableItems(items)

        return order.map { key in
            let group = groups[key] ?? []
            let primary = pickPrimaryItem(group, store: store, sourceOrder: sourceOrder)
            let variants = buildVariants(group, store: store)
            var primarySourceId = primary.sourceId
            if primarySourceId == 0, let first = variants.first {
                primarySourceId = first.sourceId
            }
            let mergedMeta = mergeMetadataFields(group, store: store, primary: primary.item)
            let latestTime = group.map(\.item.vodTime).max() ?? 0
            let stamped =
                latestTime > mergedMeta.vodTime
                ? mergedMeta.withVodTime(latestTime)
                : mergedMeta
            return MergedVodItem(
                item: stamped,
                variants: variants,
                primarySourceId: primarySourceId
            )
        }
    }

    /// Group by title; keep different concrete years apart; fold empty-year rows into the only / newest year.
    static func groupMergeableItems(
        _ items: [MergeableVodItem]
    ) -> (groups: [String: [MergeableVodItem]], order: [String]) {
        var byTitle: [String: [MergeableVodItem]] = [:]
        var titleOrder: [String] = []
        for item in items {
            let title = HomeFeed.normalizeTitle(item.item.vodName)
            guard !title.isEmpty else { continue }
            if byTitle[title] == nil {
                titleOrder.append(title)
            }
            byTitle[title, default: []].append(item)
        }

        var groups: [String: [MergeableVodItem]] = [:]
        var order: [String] = []

        for title in titleOrder {
            guard let group = byTitle[title] else { continue }
            let concreteYears = orderedUniqueYears(in: group)
            if concreteYears.count <= 1 {
                let key = concreteYears.first.map { "\(title)|\($0)" } ?? title
                order.append(key)
                groups[key] = group
                continue
            }

            var byYear: [String: [MergeableVodItem]] = [:]
            var emptyYear: [MergeableVodItem] = []
            var yearOrder: [String] = []
            for entry in group {
                let year = HomeFeed.normalizeYear(entry.item.vodYear)
                if year.isEmpty {
                    emptyYear.append(entry)
                    continue
                }
                if byYear[year] == nil {
                    yearOrder.append(year)
                }
                byYear[year, default: []].append(entry)
            }

            if !emptyYear.isEmpty {
                let targetYear = yearOrder.max { lhs, rhs in
                    let lt = byYear[lhs]?.map(\.item.vodTime).max() ?? 0
                    let rt = byYear[rhs]?.map(\.item.vodTime).max() ?? 0
                    if lt != rt { return lt < rt }
                    return false
                } ?? yearOrder[0]
                byYear[targetYear, default: []].append(contentsOf: emptyYear)
            }

            for year in yearOrder {
                let key = "\(title)|\(year)"
                order.append(key)
                groups[key] = byYear[year]
            }
        }

        return (groups, order)
    }

    private static func orderedUniqueYears(in group: [MergeableVodItem]) -> [String] {
        var seen = Set<String>()
        var years: [String] = []
        for entry in group {
            let year = HomeFeed.normalizeYear(entry.item.vodYear)
            guard !year.isEmpty, !seen.contains(year) else { continue }
            seen.insert(year)
            years.append(year)
        }
        return years
    }

    /// Release year first (future placeholders rank with update-time year), then update time.
    static func sortMergedByUpdatedDesc(_ items: [MergedVodItem]) -> [MergedVodItem] {
        items.enumerated()
            .sorted { lhs, rhs in
                let ly = HomeFeed.sortYearValue(lhs.element.item.vodYear, vodTime: lhs.element.item.vodTime)
                let ry = HomeFeed.sortYearValue(rhs.element.item.vodYear, vodTime: rhs.element.item.vodTime)
                if ly != ry {
                    return ly > ry
                }
                if lhs.element.item.vodTime != rhs.element.item.vodTime {
                    return lhs.element.item.vodTime > rhs.element.item.vodTime
                }
                return lhs.offset < rhs.offset
            }
            .map(\.element)
    }

    static func makeDetailResponse(
        source: Source,
        primary: VodItemRaw,
        extras: [PlayParser.VodWithSource],
        store: SourceStore? = nil
    ) -> DetailResponse {
        var vodsWithSource: [PlayParser.VodWithSource] = [
            PlayParser.VodWithSource(source: source, vod: primary)
        ]
        for extra in extras where !(extra.source.id == source.id && extra.vod.vodId == primary.vodId) {
            vodsWithSource.append(extra)
        }

        let playSources = PlayParser.mergePlaySources(from: vodsWithSource)
        let mergeable = vodsWithSource.map {
            MergeableVodItem(item: $0.vod, sourceId: $0.source.id, sourceName: $0.source.name)
        }
        let bestVod: VodItemRaw
        if let store, !mergeable.isEmpty {
            bestVod = mergeMetadataFields(mergeable, store: store, primary: primary)
        } else {
            bestVod = pickBestVodMetadata(vodsWithSource)
        }
        let variants = vodsWithSource
            .map { entry in
                VodVariant(sourceId: entry.source.id, sourceName: entry.source.name, vodId: entry.vod.vodId)
            }
            .sorted {
                let lhs = store?.metadataPriority(for: $0.sourceId) ?? PlayLineWeighting.sourceWeight($0.sourceId)
                let rhs = store?.metadataPriority(for: $1.sourceId) ?? PlayLineWeighting.sourceWeight($1.sourceId)
                return lhs > rhs
            }
        let vod = VodItem(
            vodId: primary.vodId,
            vodName: primary.vodName,
            vodPic: bestVod.vodPic,
            vodRemarks: bestVod.vodRemarks.isEmpty ? nil : bestVod.vodRemarks,
            vodBlurb: bestVod.vodBlurb.isEmpty ? nil : bestVod.vodBlurb,
            vodContent: bestVod.vodContent.isEmpty ? nil : bestVod.vodContent,
            typeName: bestVod.typeName.isEmpty ? nil : bestVod.typeName,
            vodClass: bestVod.vodClass.isEmpty ? nil : bestVod.vodClass,
            vodYear: bestVod.vodYear.isEmpty ? nil : bestVod.vodYear,
            variants: variants,
            primarySourceId: source.id,
            vodTime: bestVod.vodTime
        )
        return DetailResponse(
            vod: vod,
            playSources: playSources,
            variants: variants,
            merged: !extras.isEmpty
        )
    }

    static func fetchPrimaryVodDetail(source: Source, vodId: String) async throws -> (VodItemRaw, DetailResponse)? {
        let movie = try await SourceCollector.detail(source: source, sourceMovieId: vodId)
        let primary = movie.toVodItemRaw()
        return (primary, makeDetailResponse(source: source, primary: primary, extras: []))
    }

    static func enrichMergedVodDetail(
        store: SourceStore,
        source: Source,
        vodId: String,
        primary: VodItemRaw
    ) async -> DetailResponse {
        let searchable = store.collectable(capability: \.search)

        struct Match {
            let source: Source
            let vodId: String
        }

        var matches: [Match] = [Match(source: source, vodId: vodId)]
        var seen: Set<String> = ["\(source.id):\(vodId)"]

        func addMatch(_ src: Source, _ id: String) {
            let key = "\(src.id):\(id)"
            guard !seen.contains(key) else { return }
            seen.insert(key)
            matches.append(Match(source: src, vodId: id))
        }

        let searchBatches = await withTaskGroup(of: (Int, [Match]).self) { group in
            for (index, searchSource) in searchable.enumerated() {
                group.addTask {
                    guard let page = try? await SourceCollector.search(
                        source: searchSource,
                        keyword: primary.vodName,
                        page: 1
                    ) else {
                        return (index, [])
                    }
                    let found = page.list.compactMap { movie -> Match? in
                        let item = movie.toVodItemRaw()
                        guard isCompatibleVodMatch(primary, item) else { return nil }
                        return Match(source: searchSource, vodId: movie.sourceMovieId)
                    }
                    return (index, found)
                }
            }
            var buckets = Array(repeating: [Match](), count: searchable.count)
            for await (index, batch) in group {
                buckets[index] = batch
            }
            return buckets
        }

        for batch in searchBatches {
            for match in batch {
                if matches.count >= maxVariants { break }
                addMatch(match.source, match.vodId)
            }
            if matches.count >= maxVariants { break }
        }

        let detailCapable = matches.filter { $0.source.capabilities.detail }
        let extras: [PlayParser.VodWithSource] = await withTaskGroup(of: (Int, PlayParser.VodWithSource?).self) { group in
            for (index, match) in detailCapable.enumerated() {
                group.addTask {
                    if match.source.id == source.id && match.vodId == vodId {
                        return (index, PlayParser.VodWithSource(source: match.source, vod: primary))
                    }
                    guard let movie = try? await SourceCollector.detail(
                        source: match.source,
                        sourceMovieId: match.vodId
                    ) else {
                        return (index, nil)
                    }
                    let vod = movie.toVodItemRaw()
                    guard isCompatibleVodMatch(primary, vod) else { return (index, nil) }
                    return (index, PlayParser.VodWithSource(source: match.source, vod: vod))
                }
            }
            var ordered = Array<PlayParser.VodWithSource?>(repeating: nil, count: detailCapable.count)
            for await (index, entry) in group {
                ordered[index] = entry
            }
            return ordered.compactMap { $0 }
        }

        let extraOnly = extras.filter { !($0.source.id == source.id && $0.vod.vodId == primary.vodId) }
        return makeDetailResponse(source: source, primary: primary, extras: extraOnly, store: store)
    }

    static func fetchMergedVodDetail(store: SourceStore, source: Source, vodId: String) async throws -> DetailResponse? {
        guard let (primary, _) = try await fetchPrimaryVodDetail(source: source, vodId: vodId) else {
            return nil
        }
        return await enrichMergedVodDetail(store: store, source: source, vodId: vodId, primary: primary)
    }

    static func toVodItems(_ merged: [MergedVodItem]) -> [VodItem] {
        merged.map { entry in
            entry.item.toVodItem(variants: entry.variants, primarySourceId: entry.primarySourceId)
        }
    }

    static func buildVodMergeKey(_ item: VodItemRaw) -> String {
        HomeFeed.mergeKey(title: item.vodName, year: item.vodYear)
    }

    private static func countPlayLines(_ item: VodItemRaw) -> Int {
        let from = item.vodPlayFrom
        if from.isEmpty { return 0 }
        if from.contains("$$$") { return from.components(separatedBy: "$$$").filter { !$0.isEmpty }.count }
        if from.contains(",") { return from.components(separatedBy: ",").filter { !$0.isEmpty }.count }
        return 1
    }

    private static func pickPrimaryItem(
        _ items: [MergeableVodItem],
        store: SourceStore,
        sourceOrder: [Int: Int]
    ) -> MergeableVodItem {
        items.max { a, b in
            let aPic = a.item.vodPic.isEmpty ? 0 : 1
            let bPic = b.item.vodPic.isEmpty ? 0 : 1
            if aPic != bPic { return aPic < bPic }

            let aWeight = store.metadataPriority(for: a.sourceId)
            let bWeight = store.metadataPriority(for: b.sourceId)
            if aWeight != bWeight { return aWeight < bWeight }

            let aLines = countPlayLines(a.item)
            let bLines = countPlayLines(b.item)
            if aLines != bLines { return aLines < bLines }

            let aOrder = sourceOrder[a.sourceId] ?? 999
            let bOrder = sourceOrder[b.sourceId] ?? 999
            return aOrder > bOrder
        } ?? items[0]
    }

    private static func buildVariants(_ items: [MergeableVodItem], store: SourceStore) -> [VodVariant] {
        var variants: [VodVariant] = []
        var seen = Set<String>()
        for item in items where item.sourceId != 0 {
            let key = "\(item.sourceId):\(item.item.vodId)"
            guard !seen.contains(key) else { continue }
            seen.insert(key)
            variants.append(VodVariant(
                sourceId: item.sourceId,
                sourceName: item.sourceName,
                vodId: item.item.vodId
            ))
        }
        return variants.sorted {
            store.metadataPriority(for: $0.sourceId) > store.metadataPriority(for: $1.sourceId)
        }
    }

    /// Field-level merge: higher metadata_priority wins; ties broken by completeness.
    static func mergeMetadataFields(
        _ items: [MergeableVodItem],
        store: SourceStore,
        primary: VodItemRaw
    ) -> VodItemRaw {
        guard !items.isEmpty else { return primary }

        func ranked(_ pick: (VodItemRaw) -> String) -> String {
            let candidates = items
                .map { (priority: store.metadataPriority(for: $0.sourceId), value: pick($0.item), item: $0.item) }
                .filter { !$0.value.isEmpty }
            guard let best = candidates.max(by: { a, b in
                if a.priority != b.priority { return a.priority < b.priority }
                return a.value.count < b.value.count
            }) else {
                return pick(primary)
            }
            return best.value
        }

        let pic = ranked(\.vodPic)
        let remarks = ranked(\.vodRemarks)
        let year = ranked(\.vodYear)
        let area = ranked(\.vodArea)
        let vodClass = ranked(\.vodClass)
        let blurb = ranked(\.vodBlurb)
        let content = ranked(\.vodContent)
        let actor = ranked(\.vodActor)
        let director = ranked(\.vodDirector)
        let typeName = ranked(\.typeName)

        return VodItemRaw(
            vodId: primary.vodId,
            vodName: primary.vodName.isEmpty ? ranked(\.vodName) : primary.vodName,
            vodPic: pic.isEmpty ? primary.vodPic : pic,
            vodRemarks: remarks.isEmpty ? primary.vodRemarks : remarks,
            vodYear: year.isEmpty ? primary.vodYear : year,
            vodArea: area.isEmpty ? primary.vodArea : area,
            vodClass: vodClass.isEmpty ? primary.vodClass : vodClass,
            vodBlurb: blurb.isEmpty ? primary.vodBlurb : blurb,
            vodContent: content.isEmpty ? primary.vodContent : content,
            vodActor: actor.isEmpty ? primary.vodActor : actor,
            vodDirector: director.isEmpty ? primary.vodDirector : director,
            vodPlayFrom: primary.vodPlayFrom,
            vodPlayURL: primary.vodPlayURL,
            typeId: primary.typeId,
            typeName: typeName.isEmpty ? primary.typeName : typeName,
            vodTime: primary.vodTime
        )
    }

    private static func pickBestVodMetadata(_ entries: [PlayParser.VodWithSource]) -> VodItemRaw {
        entries.max { a, b in
            let aPri = a.source.priority.metadataPriority
            let bPri = b.source.priority.metadataPriority
            if aPri != bPri { return aPri < bPri }

            let aPic = a.vod.vodPic.isEmpty ? 0 : 1
            let bPic = b.vod.vodPic.isEmpty ? 0 : 1
            if aPic != bPic { return aPic < bPic }

            let aContent = a.vod.vodContent.isEmpty ? a.vod.vodBlurb.count : a.vod.vodContent.count
            let bContent = b.vod.vodContent.isEmpty ? b.vod.vodBlurb.count : b.vod.vodContent.count
            return aContent < bContent
        }?.vod ?? entries[0].vod
    }
}

File: clients/android/core/src/main/kotlin/com/heibaimiao/multilivetv/merge/VodMergeService.kt
package com.heibaimiao.multilivetv.merge

import com.heibaimiao.multilivetv.home.HomeFeed
import com.heibaimiao.multilivetv.maccms.MergeableVodItem
import com.heibaimiao.multilivetv.maccms.VodItemRaw
import com.heibaimiao.multilivetv.model.DetailResponse
import com.heibaimiao.multilivetv.model.Source
import com.heibaimiao.multilivetv.model.VodItem
import com.heibaimiao.multilivetv.model.VodVariant
import com.heibaimiao.multilivetv.parser.PlayLineWeighting
import com.heibaimiao.multilivetv.parser.PlayParser
import com.heibaimiao.multilivetv.source.SourceCapability
import com.heibaimiao.multilivetv.source.SourceCollector
import com.heibaimiao.multilivetv.source.SourceStore
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope

object VodMergeService {
    private const val MAX_VARIANTS = 16
    const val SEARCH_SOURCE_LIMIT = 6

    data class MergedVodItem(
        val item: VodItemRaw,
        val variants: List<VodVariant>,
        val primarySourceId: Int,
    )

    data class EnrichPlan(
        val known: List<VodVariant>,
        val searchSourceIds: List<Int>,
    )

    fun isCompatibleVodMatch(primary: VodItemRaw, candidate: VodItemRaw): Boolean {
        val primaryTitle = HomeFeed.titleKeyForMatch(primary.vodName)
        val candidateTitle = HomeFeed.titleKeyForMatch(candidate.vodName)
        if (primaryTitle.isEmpty() || candidateTitle.isEmpty() || primaryTitle != candidateTitle) return false
        val primaryYear = HomeFeed.normalizeYear(primary.vodYear)
            .ifEmpty { yearSuffix(primary.vodName) }
        val candidateYear = HomeFeed.normalizeYear(candidate.vodYear)
            .ifEmpty { yearSuffix(candidate.vodName) }
        if (primaryYear.isEmpty() || candidateYear.isEmpty()) return true
        return primaryYear == candidateYear
    }

    fun enrichPlan(
        primarySourceId: Int,
        primaryVodId: String,
        knownVariants: List<VodVariant>,
        searchableSourceIds: List<Int>,
        searchLimit: Int = SEARCH_SOURCE_LIMIT,
    ): EnrichPlan {
        val known = LinkedHashMap<String, VodVariant>()
        fun add(variant: VodVariant) {
            if (variant.vodId.isEmpty()) return
            known.putIfAbsent("${variant.sourceId}:${variant.vodId}", variant)
        }
        add(VodVariant(primarySourceId, "", primaryVodId))
        knownVariants.forEach(::add)
        val knownSourceIds = known.values.map { it.sourceId }.toSet()
        val search = if (knownSourceIds.size >= searchLimit) {
            emptyList()
        } else {
            searchableSourceIds.filter { it !in knownSourceIds }.take(searchLimit)
        }
        return EnrichPlan(known.values.toList(), search)
    }

    private fun yearSuffix(name: String): String {
        val match = Regex("""((?:19|20)\d{2})$""").find(HomeFeed.normalizeTitle(name)) ?: return ""
        return HomeFeed.normalizeYear(match.groupValues[1])
    }

    fun mergeVodItems(items: List<MergeableVodItem>, store: SourceStore): List<MergedVodItem> {
        val sourceOrder = store.enabled().mapIndexed { index, source -> source.id to index }.toMap()
        val (groups, order) = groupMergeableItems(items)
        return order.map { key ->
            val group = groups[key].orEmpty()
            val primary = pickPrimaryItem(group, store, sourceOrder)
            val variants = buildVariants(group, store)
            var primarySourceId = primary.sourceId
            if (primarySourceId == 0) primarySourceId = variants.firstOrNull()?.sourceId ?: 0
            val mergedMeta = mergeMetadataFields(group, store, primary.item)
            val latestTime = group.maxOfOrNull { it.item.vodTime } ?: 0
            val stamped = if (latestTime > mergedMeta.vodTime) mergedMeta.withVodTime(latestTime) else mergedMeta
            MergedVodItem(stamped, variants, primarySourceId)
        }
    }

    fun groupMergeableItems(
        items: List<MergeableVodItem>,
    ): Pair<Map<String, List<MergeableVodItem>>, List<String>> {
        val byTitle = linkedMapOf<String, MutableList<MergeableVodItem>>()
        for (item in items) {
            val title = HomeFeed.normalizeTitle(item.item.vodName)
            if (title.isEmpty()) continue
            byTitle.getOrPut(title) { mutableListOf() } += item
        }
        val groups = linkedMapOf<String, List<MergeableVodItem>>()
        val order = mutableListOf<String>()
        for ((title, group) in byTitle) {
            val concreteYears = orderedUniqueYears(group)
            if (concreteYears.size <= 1) {
                val key = concreteYears.firstOrNull()?.let { "$title|$it" } ?: title
                order += key
                groups[key] = group
                continue
            }
            val byYear = linkedMapOf<String, MutableList<MergeableVodItem>>()
            val emptyYear = mutableListOf<MergeableVodItem>()
            for (entry in group) {
                val year = HomeFeed.normalizeYear(entry.item.vodYear)
                if (year.isEmpty()) {
                    emptyYear += entry
                    continue
                }
                byYear.getOrPut(year) { mutableListOf() } += entry
            }
            if (emptyYear.isNotEmpty()) {
                val targetYear = byYear.maxByOrNull { (_, rows) -> rows.maxOf { it.item.vodTime } }?.key
                    ?: byYear.keys.first()
                byYear.getOrPut(targetYear) { mutableListOf() } += emptyYear
            }
            for ((year, rows) in byYear) {
                val key = "$title|$year"
                order += key
                groups[key] = rows
            }
        }
        return groups to order
    }

    fun sortMergedByUpdatedDesc(items: List<MergedVodItem>): List<MergedVodItem> =
        items.mapIndexed { index, item -> index to item }
            .sortedWith(
                compareByDescending<Pair<Int, MergedVodItem>> {
                    HomeFeed.sortYearValue(it.second.item.vodYear, it.second.item.vodTime)
                }
                    .thenByDescending { it.second.item.vodTime }
                    .thenBy { it.first },
            )
            .map { it.second }

    fun makeDetailResponse(
        source: Source,
        primary: VodItemRaw,
        extras: List<PlayParser.VodWithSource>,
        store: SourceStore? = null,
    ): DetailResponse {
        val vodsWithSource = mutableListOf(PlayParser.VodWithSource(source, primary))
        for (extra in extras) {
            if (extra.source.id == source.id && extra.vod.vodId == primary.vodId) continue
            vodsWithSource += extra
        }
        val playSources = PlayParser.mergePlaySources(vodsWithSource)
        val mergeable = vodsWithSource.map {
            MergeableVodItem(it.vod, it.source.id, it.source.name)
        }
        val bestVod = if (store != null && mergeable.isNotEmpty()) {
            mergeMetadataFields(mergeable, store, primary)
        } else {
            pickBestVodMetadata(vodsWithSource)
        }
        val variants = vodsWithSource
            .map { VodVariant(it.source.id, it.source.name, it.vod.vodId) }
            .sortedByDescending {
                store?.metadataPriority(it.sourceId) ?: PlayLineWeighting.sourceWeight(it.sourceId)
            }
        val vod = VodItem(
            vodId = primary.vodId,
            vodName = primary.vodName,
            vodPic = bestVod.vodPic,
            vodRemarks = bestVod.vodRemarks.ifEmpty { null },
            vodBlurb = bestVod.vodBlurb.ifEmpty { null },
            vodContent = bestVod.vodContent.ifEmpty { null },
            typeName = bestVod.typeName.ifEmpty { null },
            vodClass = bestVod.vodClass.ifEmpty { null },
            vodYear = bestVod.vodYear.ifEmpty { null },
            variants = variants,
            primarySourceId = source.id,
            vodTime = bestVod.vodTime,
        )
        return DetailResponse(vod, playSources, variants, merged = extras.isNotEmpty())
    }

    suspend fun fetchPrimaryVodDetail(source: Source, vodId: String): Pair<VodItemRaw, DetailResponse>? {
        val movie = SourceCollector.detail(source, vodId)
        val primary = movie.toVodItemRaw()
        return primary to makeDetailResponse(source, primary, extras = emptyList())
    }

    suspend fun enrichMergedVodDetail(
        store: SourceStore,
        source: Source,
        vodId: String,
        primary: VodItemRaw,
        knownVariants: List<VodVariant> = emptyList(),
    ): DetailResponse = coroutineScope {
        val searchable = store.collectable(SourceCapability.SEARCH)
            .sortedByDescending { store.metadataPriority(it.id) + store.playPriority(it.id) }
        val plan = enrichPlan(source.id, vodId, knownVariants, searchable.map { it.id })
        data class Match(val source: Source, val vodId: String)
        val matches = mutableListOf<Match>()
        val seen = mutableSetOf<String>()
        fun addMatch(match: Match) {
            if (matches.size >= MAX_VARIANTS) return
            val key = "${match.source.id}:${match.vodId}"
            if (!seen.add(key)) return
            matches += match
        }
        for (variant in plan.known) {
            val src = store.byID(variant.sourceId) ?: continue
            if (!src.capabilities.detail) continue
            addMatch(Match(src, variant.vodId))
        }
        val keyword = HomeFeed.searchKeyword(primary.vodName)
        val searchSources = plan.searchSourceIds.mapNotNull { store.byID(it) }
        if (searchSources.isNotEmpty()) {
            val searchBatches = searchSources.mapIndexed { index, searchSource ->
                async {
                    val page = runCatching { SourceCollector.search(searchSource, keyword, 1) }.getOrNull()
                    val found = page?.list.orEmpty().mapNotNull { movie ->
                        val item = movie.toVodItemRaw()
                        if (!isCompatibleVodMatch(primary, item)) null
                        else Match(searchSource, movie.sourceMovieId)
                    }
                    index to found
                }
            }.awaitAll().sortedBy { it.first }.map { it.second }
            for (batch in searchBatches) {
                for (match in batch) {
                    addMatch(match)
                    if (matches.size >= MAX_VARIANTS) break
                }
                if (matches.size >= MAX_VARIANTS) break
            }
        }
        val extras = matches.mapIndexed { index, match ->
            async {
                if (match.source.id == source.id && match.vodId == vodId) {
                    index to PlayParser.VodWithSource(match.source, primary)
                } else {
                    val movie = runCatching { SourceCollector.detail(match.source, match.vodId) }.getOrNull()
                    val vod = movie?.toVodItemRaw()
                    if (vod != null && isCompatibleVodMatch(primary, vod)) {
                        index to PlayParser.VodWithSource(match.source, vod)
                    } else {
                        index to null
                    }
                }
            }
        }.awaitAll().sortedBy { it.first }.mapNotNull { it.second }
        val extraOnly = extras.filter { !(it.source.id == source.id && it.vod.vodId == primary.vodId) }
        makeDetailResponse(source, primary, extraOnly, store)
    }

    fun toVodItems(merged: List<MergedVodItem>): List<VodItem> =
        merged.map { it.item.toVodItem(it.variants, it.primarySourceId) }

    fun mergeMetadataFields(
        items: List<MergeableVodItem>,
        store: SourceStore,
        primary: VodItemRaw,
    ): VodItemRaw {
        if (items.isEmpty()) return primary
        fun ranked(pick: (VodItemRaw) -> String): String {
            val candidates = items.map {
                Triple(store.metadataPriority(it.sourceId), pick(it.item), it.item)
            }.filter { it.second.isNotEmpty() }
            val best = candidates.maxWithOrNull(
                compareBy<Triple<Int, String, VodItemRaw>> { it.first }.thenBy { it.second.length },
            )
            return best?.second ?: pick(primary)
        }
        val pic = ranked { it.vodPic }
        val remarks = ranked { it.vodRemarks }
        val year = ranked { it.vodYear }
        val area = ranked { it.vodArea }
        val vodClass = ranked { it.vodClass }
        val blurb = ranked { it.vodBlurb }
        val content = ranked { it.vodContent }
        val actor = ranked { it.vodActor }
        val director = ranked { it.vodDirector }
        val typeName = ranked { it.typeName }
        return VodItemRaw(
            vodId = primary.vodId,
            vodName = primary.vodName.ifEmpty { ranked { it.vodName } },
            vodPic = pic.ifEmpty { primary.vodPic },
            vodRemarks = remarks.ifEmpty { primary.vodRemarks },
            vodYear = year.ifEmpty { primary.vodYear },
            vodArea = area.ifEmpty { primary.vodArea },
            vodClass = vodClass.ifEmpty { primary.vodClass },
            vodBlurb = blurb.ifEmpty { primary.vodBlurb },
            vodContent = content.ifEmpty { primary.vodContent },
            vodActor = actor.ifEmpty { primary.vodActor },
            vodDirector = director.ifEmpty { primary.vodDirector },
            vodPlayFrom = primary.vodPlayFrom,
            vodPlayURL = primary.vodPlayURL,
            typeId = primary.typeId,
            typeName = typeName.ifEmpty { primary.typeName },
            vodTime = primary.vodTime,
        )
    }

    private fun orderedUniqueYears(group: List<MergeableVodItem>): List<String> {
        val years = mutableListOf<String>()
        val seen = mutableSetOf<String>()
        for (entry in group) {
            val year = HomeFeed.normalizeYear(entry.item.vodYear)
            if (year.isEmpty() || !seen.add(year)) continue
            years += year
        }
        return years
    }

    private fun countPlayLines(item: VodItemRaw): Int {
        val from = item.vodPlayFrom
        if (from.isEmpty()) return 0
        if (from.contains("$$$")) return from.split("$$$").count { it.isNotEmpty() }
        if (from.contains(",")) return from.split(",").count { it.isNotEmpty() }
        return 1
    }

    private fun pickPrimaryItem(
        items: List<MergeableVodItem>,
        store: SourceStore,
        sourceOrder: Map<Int, Int>,
    ): MergeableVodItem = items.maxWith(
        compareBy<MergeableVodItem> { if (it.item.vodPic.isEmpty()) 0 else 1 }
            .thenBy { store.metadataPriority(it.sourceId) }
            .thenBy { countPlayLines(it.item) }
            .thenByDescending { sourceOrder[it.sourceId] ?: 999 }
            .reversed()
            .thenBy { sourceOrder[it.sourceId] ?: 999 },
    )

    private fun buildVariants(items: List<MergeableVodItem>, store: SourceStore): List<VodVariant> {
        val variants = mutableListOf<VodVariant>()
        val seen = mutableSetOf<String>()
        for (item in items) {
            if (item.sourceId == 0) continue
            val key = "${item.sourceId}:${item.item.vodId}"
            if (!seen.add(key)) continue
            variants += VodVariant(item.sourceId, item.sourceName, item.item.vodId)
        }
        return variants.sortedByDescending { store.metadataPriority(it.sourceId) }
    }

    private fun pickBestVodMetadata(entries: List<PlayParser.VodWithSource>): VodItemRaw =
        entries.maxWith(
            compareBy<PlayParser.VodWithSource> { it.source.priority.metadataPriority }
                .thenBy { if (it.vod.vodPic.isEmpty()) 0 else 1 }
                .thenBy { if (it.vod.vodContent.isEmpty()) it.vod.vodBlurb.length else it.vod.vodContent.length },
        ).vod
}
